### PR TITLE
add multi-stage loss

### PR DESCRIPTION
**Pull request by wusize**
add multi-stage loss

### CODE REVIEW

File: mmpose/models/detectors/top_down.py [hunk 1]
```diff
@@ -115,30 +115,61 @@ def forward_train(self, img, target, target_weight, img_metas, **kwargs):
...
+                    target[:, -1, :, :, :][target_weight[:, -1, :, :].squeeze(
+                        -1) > 0].unsqueeze(0).detach().cpu().numpy())
+            else:
+                _, avg_acc, cnt = pose_pck_accuracy(
+                    output[-1][target_weight.squeeze(-1) > 0].unsqueeze(
+                        0).detach().cpu().numpy(),
+                    target[target_weight.squeeze(-1) > 0].unsqueeze(
+                        0).detach().cpu().numpy())
         else:
             _, avg_acc, cnt = pose_pck_accuracy(
```
Comment (innerlee) [inline mmpose/models/detectors/top_down.py:168]: are these `cnt`s used?

File: mmpose/models/detectors/top_down.py [hunk 2]
```diff
@@ -115,30 +115,61 @@ def forward_train(self, img, target, target_weight, img_metas, **kwargs):
...
 
         if isinstance(output, list):
-            _, avg_acc, cnt = pose_pck_accuracy(
-                output[-1][target_weight.squeeze(-1) > 0].unsqueeze(
-                    0).detach().cpu().numpy(),
-                target[target_weight.squeeze(-1) > 0].unsqueeze(
-                    0).detach().cpu().numpy())
+            if target.dim() == 5 and target_weight.dim() == 4:
+                _, avg_acc, cnt = pose_pck_accuracy(
+                    output[-1][target_weight[:, -1, :, :].squeeze(-1) > 0].
```
Comment (innerlee) [inline mmpose/models/detectors/top_down.py:157]: question: why is the last output special

File: mmpose/models/detectors/top_down.py [hunk 3]
```diff
@@ -188,7 +218,7 @@ def forward_test(self, img, img_metas, **kwargs):
             score = np.array(img_metas['bbox_score']).reshape(-1)
 
         preds, maxvals = keypoints_from_heatmaps(
-            output_heatmap,
+            output.clone().cpu().numpy(),
```
Comment (innerlee) [inline mmpose/models/detectors/top_down.py:221]: wrong change?

File: mmpose/models/detectors/top_down.py [hunk 4]
```diff
@@ -207,7 +237,7 @@ def forward_test(self, img, img_metas, **kwargs):
         all_boxes[0, 5] = score
         image_path.extend(img_metas['image_file'])
 
-        return all_preds, all_boxes, image_path, output_heatmap
```
Comment (innerlee) [inline mmpose/models/detectors/top_down.py:210]: i guess something went wrong during rebase

**Comment by innerlee** [comment]
the main change is ok, but there are some unintended changes

File: tests/test_model/test_top_down_forward.py [hunk 5]
```diff
@@ -84,8 +84,62 @@ def test_topdown_forward():
...
+        ),
+        keypoint_head=dict(
+            type='TopDownMultiStageHead',
+            in_channels=256,
+            out_channels=17,
+            num_stages=1,
+            num_deconv_layers=0,
+            extra=dict(final_conv_kernel=1, ),
+        ),
+        # keypoint_head=dict(
```
Comment (jin-s13) [inline tests/test_model/test_top_down_forward.py:102]: del the unnecessary comments

File: mmpose/models/detectors/top_down.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/models/detectors/top_down.py:157]: Generally speaking, the last output is the final prediction (used for calculate accuracy), while the intermediate outputs are used for training.

Comment (innerlee) [reply mmpose/models/detectors/top_down.py:157]: cool, @wusize may add a line of comment explaining this magic number

File: mmpose/models/detectors/top_down.py (hunk 1 above)
Comment (wusize) [reply mmpose/models/detectors/top_down.py:168]: Nope, I think. These cnts have been here in previous versions. Should I change the cnts into _ ?

Comment (innerlee) [reply mmpose/models/detectors/top_down.py:168]: yeah

**Review by innerlee — APPROVED** [review]

**Comment by OpenMMLab-Team** [comment]
Hi @wusize First of all, we want to express our gratitude for your significant PR in the MMpose project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
